torch_xla/experimental/apply_layers.py

```diff
@@ -0,0 +1,72 @@
+from typing import Iterable
+
+import torch
+import torch.nn as nn
+from torch.utils._pytree import tree_map
+
+from torch_xla.experimental.scan import scan
+
+
+def apply_layers(layers: Iterable[torch.nn.Module], input_data):
+  """Applies each layer in `layers` to `input_data` sequentially.
+
+  `input_data` is provided as input to the first layer in `layers`. The output of one
+  layer is provided as input to next layer. This function is equivalent to
+
+    sequential = torch.nn.Sequential(layers)
+    sequential(input_data)
+
+  This function can be faster to compile since it reuses the XLA computation of the
+  first layer to perform the computation of all other layers.
+  """
+  # Handle empty layers case.
+  try:
+    next(iter(layers))
+  except StopIteration:
+    return input_data
+
+  # Extract and stack the parameters into a pytree.
+  params = [_extract_weights_dict(layer) for layer in layers]
```

Comment:
what if it is a dropout layer that parameters are more than just tensors?

Reply:
If there is a dropout layer that references other parameters (for example, the dropout probability), then those tensors will be captured as an additional input to the HLO graph. As implemented now, `apply_layers` and `scan` will trace the first layer, and use the same captured tensor for subsequent layers. This will be a problem if the user passes different dropout probabilities for say a sequence of dropout layers -- we'll instead incorrectly just keep using the first dropout's probability. I'll have to dig deeper and find a solution for this.

Reply:
This is now handled correctly now that I'm splitting the RNG for each layer.